src/Orleans/Lifecycle/LifecycleObservable.cs

```diff
@@ -0,0 +1,125 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Orleans.Runtime;
+
+namespace Orleans
+{
+    public class LifecycleObservable<TStage> : ILifecycleObservable<TStage>, ILifecycleObserver
+    {
+        private readonly ConcurrentDictionary<object, OrderedObserver> subscribers;
+        private readonly Logger logger;
+        private TStage highStage;
+
+        public LifecycleObservable(Logger logger)
+        {
+            this.logger = logger?.GetLogger(GetType().Name);
+            subscribers = new ConcurrentDictionary<object, OrderedObserver>();
```

Comment:
Maybe NIT: since you are creating a new file, I think we should use the keyword `this` each time you are using a private/protected class member. I know here you did it because of the ambiguous `logger` but further in the code you used `this`